### PR TITLE
bootstrap logic in server side

### PR DESCRIPTION
**Pull request by voyager-dw**
This PR is for bootstrap logic in server side, will add more test and add some more logic

### CODE REVIEW

**Comment by voyager-dw** [comment]
items to add:
tests
buffer io test together with chunk size 
purge enable/disable
session recycle
topology retrieve to verify

File: datanode/bootstrap/bootstrap_server.go [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	vp := &pb.VectorPartyRawData{}
+	buf := make([]byte, chunkSize)
+	done := false
+	for !done {
+		n, err := reader.Read(vp.Chunk)
```
Comment (jshencode) [inline datanode/bootstrap/bootstrap_server.go]: Read(buf)?

File: datanode/bootstrap/bootstrap_server.go [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+		if err != nil {
+			if err == io.EOF {
+				done = true
```
Comment (jshencode) [inline datanode/bootstrap/bootstrap_server.go]: nit: seems we don't need this done variable

File: datanode/bootstrap/bootstrap_server.go [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+		}
+	}
+	// in milsec
+	timeElapsed = utils.Now().Sub(timeStart).Nanoseconds() / 1000000
+
+	return nil
+}
+
+func (p *PeerDataNodeServerImpl) validateSessionSource(sessionID int64, nodeID string) error {
+	if sessionID == 0 {
```
Comment (jshencode) [inline datanode/bootstrap/bootstrap_server.go]: if it is sessionID is int63(),  it can be 0 right?

File: datanode/bootstrap/bootstrap_server.go [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+// StartSession create new session for one table/shard/node, only One session can be established on one table/shard from one node
+func (p *PeerDataNodeServerImpl) StartSession(ctx context.Context, req *pb.StartSessionRequest) (*pb.Session, error) {
+	var err error
+	sessionInfo := &sessionInfo{
+		table:        req.Table,
+		shardID:      req.Shard,
+		nodeID:       req.NodeID,
+		ttl:          req.Ttl,
+		lastLiveTime: utils.Now(),
+		sessionID:    rand.Int63(),
```
Comment (jshencode) [inline datanode/bootstrap/bootstrap_server.go]: monotonically increasing number might be better for the id?

File: datanode/bootstrap/bootstrap_server.go (hunk 1 above)
Comment (voyager-dw) [reply datanode/bootstrap/bootstrap_server.go]: yeah, fixed

File: datanode/bootstrap/bootstrap_server.go (hunk 4 above)
Comment (shz117) [reply datanode/bootstrap/bootstrap_server.go]: how about some combination of table shard node and sequence number

Comment (voyager-dw) [reply datanode/bootstrap/bootstrap_server.go]: hmm, that works too local id generator, will change

File: datanode/bootstrap/bootstrap_server.go (hunk 3 above)
Comment (voyager-dw) [reply datanode/bootstrap/bootstrap_server.go]: yeah, will change

File: datanode/bootstrap/bootstrap_server.go [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+		}
+		if n > 0 {
+			vp.Chunk = buf[:n]
+			if err = stream.Send(vp); err != nil {
+				return err
+			}
+		}
+	}
+	// in milsec
+	timeElapsed = utils.Now().Sub(timeStart).Nanoseconds() / 1000000
```
Comment (shz117) [inline datanode/bootstrap/bootstrap_server.go]: 1000000 --> time.Millisecond

File: datanode/bootstrap/bootstrap_server.go [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+		tableShardSessions: make(map[tableShardPair][]int64),
+	}
+}
+
+// IsBootstrapRunning is to check if any bootstrap is running in the table/shard
+func (p *PeerDataNodeServerImpl) IsBootstrapRunning(tableName string, shardID uint32) bool {
+	p.RLock()
+	defer p.RUnlock()
+
+	_, ok := p.tableShardSessions[tableShardPair{table: tableName, shardID: shardID}]
```
[the remark this answers is not in the document]
Comment (voyager-dw) [reply datanode/bootstrap/bootstrap_server.go]: yeah, that part not done yet

**Pull request merged**